### PR TITLE
Adds additional validators and "write_script" routine

### PR DESCRIPTION
**Pull request by DrPaulSharp**

### CODE REVIEW

File: tests/test_project.py [hunk 1]
```diff
@@ -566,6 +633,16 @@ def test_get_contrast_model_field(input_calc: 'RAT.project.CalcTypes', input_mod
...
+def test_write_script(test_project, test_project_script) -> None:
+    """Test the script we write to regenerate the project is as expected."""
+    mock_file = mock.mock_open()
+    with mock.patch('builtins.open', mock_file):
+        test_project.write_script()
+
+    # We check the calls to the write function as we do not return what we write
+    assert mock_file().write.call_args_list == test_project_script
+
+
```
Comment (DrPaulSharp) [inline tests/test_project.py:645]: After doing this, is there a way we can run the script we have and compare the resulting object against "test_project". I've tried to put the script together as a string and use exec, but I can't return the object that way.

File: RAT/project.py [hunk 2]
```diff
@@ -196,12 +198,20 @@ def model_post_init(self, __context: Any) -> None:
...
-        self.parameters.insert(0, RAT.models.ProtectedParameter(name='Substrate Roughness', min=1, value=3, max=5,
-                                                                fit=True, prior_type=RAT.models.Priors.Uniform, mu=0,
-                                                                sigma=np.inf))
+        if 'Substrate Roughness' not in self.parameters.get_names():
+            self.parameters.insert(0, RAT.models.ProtectedParameter(name='Substrate Roughness', min=1.0, value=3.0,
+                                                                    max=5.0, fit=True,
+                                                                    prior_type=RAT.models.Priors.Uniform, mu=0.0,
+                                                                    sigma=np.inf))
+        elif 'Substrate Roughness' not in self.get_all_protected_parameters().values():
+            # If substrate roughness is included as a standard parameter replace it with a protected parameter
```
[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RAT/project.py:207]: That's strange, it does show up in my coverage.

[the remark this answers is not in the document]
Comment (StephenNneji) [reply RAT/project.py:207]: Screenshot 2023-10-19 124438

[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RAT/project.py:207]: image
I wasn't doubting you at all - something strange about the windows test then.

[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply RAT/project.py:207]: (Those two lines are elsewhere)

File: tests/test_project.py [hunk 3]
```diff
@@ -566,6 +633,16 @@ def test_get_contrast_model_field(input_calc: 'RAT.project.CalcTypes', input_mod
     assert project.get_contrast_model_field() == expected_field_name
 
 
+def test_write_script(test_project, test_project_script) -> None:
+    """Test the script we write to regenerate the project is as expected."""
+    mock_file = mock.mock_open()
```
[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply tests/test_project.py:638]: Hmm, looking at `mkstemp()` it seems to me that you have to declare when when creating the tempfile where it is written to, so testing that wouldn't be helpful. I'm not sure either approach to this test has a clear advantage over the other really.

File: tests/test_project.py [hunk 4]
```diff
@@ -566,6 +649,47 @@ def test_get_contrast_model_field(input_calc: 'RAT.project.CalcTypes', input_mod
     assert project.get_contrast_model_field() == expected_field_name
 
 
+@pytest.mark.parametrize("input_filename", [
+    "test_script.py",
+    "test_script"
+])
+def test_write_script(test_project, test_project_script, input_filename: str) -> None:
+    """Test the script we write to regenerate the project is created and runs as expected."""
+    path = tempfile.mkdtemp(dir='.')
```
[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply tests/test_project.py]: Ok, now done. Although avoiding all of this was why I had a mock file in the first place.

**Pull request merged**